lib/bootstrap/bundler.rb

```diff
@@ -196,6 +196,41 @@ def debug?
       ENV["DEBUG"]
     end
 
+    # @param plugin_names [Array] logstash plugin names that are going to update
+    # @return [Array] gem names that plugins depend on, including logstash plugins
+    def consolidate_dependencies(plugin_names)
+      plugin_names = Array(plugin_names) if plugin_names.is_a?(String)
+
+      # get gem names in Gemfile.lock. If file doesn't exist, it will be generated
+      lockfile_gems = ::Bundler::definition.specs.to_a.map { |stub_spec| stub_spec.name }.to_set
+
+      # get the array of dependencies which are eligible to update. Bundler unlock these gems in update process
+      # exclude the gems which are not in lock file. They should not be part of unlock gems.
+      # The core libs are not expected to update when user do plugins update
+      unlock_libs = plugin_names.map { |plugin_name| fetch_plugin_dependencies(plugin_name) }
+                                .flatten.uniq
+                                .select { |lib_name| lockfile_gems.include?(lib_name) }
```

Comment:
I think we need to safeguard this to _also_ only select plugins that the Logstash team owns and expects to be API-compatible (e.g., `logstash-mixin-XYZ` whose API version is 1.x?).

Reply:
agree that we can limit to Logstash own gems. For API version compatibility, bundler is doing a good job to consider dependencies from all other gems in lock file. Once `logstash-mixin-*` is unlocked, we can trust bundler to pick the best version.  Let me know if you think we should fix API version here.